core/src/AzureMcp.Core/Services/Telemetry/UnixMachineInformationProvider.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Text;
+using Microsoft.Extensions.Logging;
+
+namespace AzureMcp.Core.Services.Telemetry;
+
+internal abstract class UnixMachineInformationProvider(ILogger<UnixMachineInformationProvider> logger)
+    : MachineInformationProviderBase(logger)
+{
+    private readonly ILogger<UnixMachineInformationProvider> _logger = logger;
+
+    /// <summary>
+    /// Gets the root folder to cache information to.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">If there is no folder to persist data in.</exception>
+    public abstract string GetStoragePath();
+
+    public override async Task<string?> GetOrCreateDeviceId()
+    {
+        string cachePath;
+        try
+        {
+            cachePath = Path.Combine(GetStoragePath(), MicrosoftDirectory, DeveloperToolsDirectory);
+        }
+        catch (InvalidOperationException ex)
+        {
+            _logger.LogWarning(ex, "Unable to find folder to cache device id to.");
+            return null;
+        }
+
+        var existingValue = await ReadValueFromDisk(cachePath, DeviceId);
+        if (existingValue != null)
+        {
+            return existingValue;
+        }
+
+        var deviceId = GenerateDeviceId();
+        if (await WriteValueToDisk(cachePath, DeviceId, deviceId))
+        {
+            return deviceId;
+        }
+        else
+        {
+            _logger.LogWarning("Unable to persist deviceId.");
+            return null;
+        }
+    }
+
+    /// <summary>
+    /// Try and write the value to disk. If <paramref name="value"/> is null or empty, this method will return false.
+    /// </summary>
+    /// <param name="directoryPath">Directory path to write the file.</param>
+    /// <param name="fileName">The name of the file.</param>
+    /// <param name="value">The value to write in the file.</param>
+    /// <returns>True, if the value was successfully written.</returns>
+    /// 
+    public async virtual Task<bool> WriteValueToDisk(string directoryPath, string fileName, string? value)
+    {
+        // If the value is not set, return immediately.
+        if (string.IsNullOrWhiteSpace(value))
+        {
+            return false;
+        }
+
+        try
+        {
+            // Create the directory if it doesn't exist.
+            if (!Directory.Exists(directoryPath))
+            {
+                Directory.CreateDirectory(directoryPath);
+            }
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, "Unable to create directory. {Directory}", directoryPath);
+            return false;
+        }
+
+        var fullPath = Path.Combine(directoryPath, fileName);
+
+        try
+        {
+            File.Delete(fullPath);
+
+            await File.WriteAllTextAsync(fullPath, value, Encoding.UTF8);
```

Comment:
`File.WriteAllTextAsync` will truncate and overwrite the file, do we need to call delete beforehand?

Reply:
When I was looking at the documentation, what it ended up doing was appending the text into the file instead of overwriting it.